src/Servant/MatrixParam/Internal.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE DeriveDataTypeable     #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs                  #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE PolyKinds              #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeOperators          #-}
+{-# LANGUAGE UndecidableInstances   #-}
+
+{-# OPTIONS_GHC -fno-warn-incomplete-patterns #-}
+module Servant.MatrixParam.Internal where
+
+import           Data.List       (foldl')
+import           Data.Map.Strict
+import qualified Data.Map.Strict as M
+import           Data.Maybe
+import           Data.Monoid
+import           Data.Proxy      (Proxy (..))
+import qualified Data.Text       as T
+import           Data.Typeable   (Typeable)
+import           GHC.TypeLits    (KnownSymbol, Symbol, symbolVal)
+import           Web.HttpApiData
+
+import Servant.API (Capture)
+
+
+-- | Expresses matrix parameters for path segments in APIs, e.g.
+--
+-- @/books;author=<author name>@
+--
+-- would be represented as:
+--
+-- >>> import Servant
+-- >>>
+-- >>> type MyApi = WithMatrixParams "book" '[MatrixParam "author" String] :> Get '[JSON] [String]
+type WithMatrixParams path params = Capture "matrixParamCapture" (MatrixParams path params)
+
+data MatrixParams (path :: Symbol) (paramSpecs :: [*]) where
+  MatrixEmpty :: MatrixParams path '[]
+  (:.:) :: Maybe a -> MatrixParams path as -> MatrixParams path (MatrixParam seg a ': as)
+    deriving (Typeable)
+infixr 3 :.:
+
+class Unmatrix ls a | ls -> a where
```

Comment:
This seems to be convenience functionality and should be clearly separated from the essential concepts, I think.